### ISSUE TITLE
Inferred recursive types

### ISSUE THREAD

**Issue by m93a** [colinhacks/zod#1689]
I'm working on a library that needs type definitions at runtime, and I wanted to use Zod as it's the _“lingua franca”_ of runtime type definitions. However, my library will make heavy use of recursive types, and working with these is currently very cumbersome in Zod. The [Recursive types section](https://zod.dev/?id=recursive-types) of the documentation reads:
> You can define a recursive schema in Zod, but because of a limitation of TypeScript, their type can't be statically inferred. Instead you'll need to define the type definition manually, and provide it to Zod as a "type hint".
> Unfortunately this code is a bit duplicative, since you're declaring the types twice: once in the interface and again in the Zod definition.

I wonder if such a limitation is necessary, since TypeScript supports recursion even on anonymous types. For example, if we define the following helper types...
```typescript
const TAG = Symbol('TAG');
type Tag<N extends string> = { [TAG]: N };
type RecursiveType<Name extends string, Self extends object, O = Self>
    = O extends Tag<Name> ? RecursiveType<Name, Self>
    : O extends Array<infer V> ? Array<RecursiveType<Name, Self, V>>
    : O extends object ? { [K in keyof O]: RecursiveType<Name, Self, O[K]> }
    : O;
```
... we can create the `Category` type without having its definition reference itself:
```typescript
type Category = RecursiveType<'Cat', {
  name: string;
  subcategories: Tag<'Cat'>[];
}>;

const x: Category = { name: "foo", subcategories: [42] }; // error
const y: Category = {
  name: "bar",
  subcategories: [
    { name: "qux", subcategories: [] }
  ]
}; // ok
```
[Link to TS Playground.](https://www.typescriptlang.org/play?#code/FAehAIAsFMBsAdoCcDOxgGMD2A7FAXcAFQEEBxcAXnAGUBPAWwCMtYAKAclLI4EoBuYPjqJiAQwDmAHgBy4aAA980HABMU4AkgCWOCQD4q4AN7gA2twC6ALnByAvoOGiAStAwBXVNoBu0IiLQsmIM0PJKKuqa+Dp6ADS0cABm4cpqGlhMAFbu+AkA8kY0yfrA4OVGhYppUUSSwaGGAPzgbp7efgGIDdAJxbBJpRXgtlUR6eAkSEhidFK6ScjgAGrNk9OzUm1eKL7+gT19yQmrQxWjqZEZ2bngLaZmANLguuAA1tB0WCn5Nq3uOz2XSCMhCvUSAwKT0shnsZXO4HygnQYHASABHTCzjCqmgSV02nw2lw6Gx4AAwmJlBIsEg6EZtpjgVIOJT8BwEsZ4TgwbYtLoJIJyigPEwMFToDSdNAULY6tJWVSOPozJZBPZ9MjQBBFCF4LAZehsHhCApbGzJbT6dRTDzQrYAERJLBYB0JEViiVS7Qy2xmAAsACZLOBHOBUcgkLTMLgCOA6OavVajFzynboI6mGIkG74R7xdTaT7ZeZ4eVbbzwA6AI4eBRuzSiguW6Ul1Wh+GWYBh1FYN7AIA)

It is then hard to imagine that Zod couldn't support recursive types with full static inference using an API like this:
```typescript
const Category = z.label('Cat', z.object({
  name: z.string(),
  subcategories: z.ref('Cat').array(),
}));
```

Is there a good reason why this isn't a part of Zod already? Or should I try to make a PR for this?

----

**EDIT:** Changed the last code snippet, so that it uses methods `z.label` and `z.ref` instead of `z.recursiveType` and `z.tag`, becaue I think the new methods are easier to understand.

**Comment by colinhacks** [comment]
In your examples, the type is not being inferred or extracted from a self-referencing runtime construct. It's trivially easy to write a recursive type. I didn't fully follow your code but this is all you actually need:

```ts
type Cat = {
  name: string;
  subcategories: Cat[];
};
```

I'm very dissatisfied with this limitation of Zod though, and overcoming this will likely be a major design goal of a hypothetic Zod 4. Anyway - you're certainly welcome to attempt a PR but I'm pretty confident this is not possible atm.

**Comment by m93a** [comment]
I'm sorry for not being clear enough. What I wanted to say was that **it _is_ possible to declare a self-referencing schema whose TS type is correctly inferred**.

While TS can't infer if from the `z.lazy` pattern, it works if we introduce the `label-ref` pattern that I propose.
[**See my minimal implementation on the TS playground.**](https://www.typescriptlang.org/play?ts=4.9.4#code/LAKA9GAEDuD2BOBrAzgQlAY1gO2QF0gGEBDPAUwHMEBPSAXkgC8A6AG2ICMzWAKAchJ4+AGibNYHAFZkMeHgG9QkZZGzEAtmQBcY-PACW2CjwCUwpSuQBXDhlKUE+ssh0t4ZAGb9BfE82Lw8MTUpuYgAL4mJgDcoHEgoBCQeAAWZMnUAA7p+siQ7sSsrLSp6e4YVvDI+gBu6Thk6CBYuAQYrsyGHmTwADx4WWSwHkT2VPDUAHz0kIogKqoa2pAARIiwsKyQiGM0K2EL1ra7Bs46ANoWC8ryi5o6K0d2ePuQTydOLpDnALqQ4Qdrjc7ssVh4NhwAq93uRxp8dAAWABM-1ESR68AQqEgPGIeX0BGQKVgVlYABNIFwTFdID8wuFYgkmfEWcy2YkoNAUqRIOCirBoPi8sRIOpDPp1IVcnh9Bh8mQJZlWGRNNg8KR9DhIMM3mQ6kEtgAtWAUzSpE3IDmQPEwbhbG2lSCZTGZWDIMgUljsLj27Ce5juEZm4lky1MskydjuEHITLEDDpRizGmGcjwDzx9KEWDqdQ4ACyZHNoeT8yB1sCwVMOgAgpXqABlDBpSX9FK5SaM8vavoAFUgZAAHuQ-XkWE2W8RfpMeMw57BSvAdL2TDoAKrYTXYCcq4i9c6pXKiOfMXs-Ts0hbEP29ftDkclneS6ez+eL5eryAASTVPXdsi3J89wPdtkGPOczwvMtlHCeIFiHV14AIVMegzBNIAbPADCMICB2HMhRyIHM82wQtizyOZuwGbIdD4PRDAoPguxUWDoLwxC2hwfA3iwhidFMehpkw7CKCAxkaQQhBkN-dNM0gAA5Kx1C4eBcPvAiS2zXMCyLEMKMvFRqOWPhsCUlSmJpVj4MHDjIBabjTOUnp+JMQSFLMnoxLglRJKQyAUNk9CACF9AoVM1PwwitJIsi9NLKjBlojhQtTCy2KsnybKkuyuIIZKwrVFy3JCgq8C8plrNsgK0PSIKNmVa8IofPJop08j4vLIykvqshrzShYMuUXzONaSkeuvIq6GmOrNl67dm13cS2OG-yZJqjDqGUzYmo0lriLauLKM6xLIDozaOE2fqWIkrK-Pswlzs2SahMe1hyppIzIAABQMMUZTqGZ6KMSAAB9VA8+BQbeV6ofy1NYfG7AodMoooasP1PEMD1mKG27pLTdaABkCR6QogN6AA1PDmu+36CVqMhpnUqL9tI3SLQ6oEutO1gSYNK6gRqQorGWCmcf+G7bPuyBebTQodFvandtpiV6bqGchdYEWPzc4m5behbW17KDJey6q5LrIJG0NvdjaVlntLZ9qjq5k6+ACK2BeubgVQIvBl3Fwb2Oy6WPeCBXjZ4H3VX9yAVzcy3gnJ42lsqs21rkjdAJtxXmZLa9qGne3NNZ2KOZd65ub4dGty9hZYEyGUuIDyzTbu3LIBrnAI+L4VsEL89X3ERuty+eOpsgLOcGTk3lrx1aCbkn803-Jv5snXPIvz-ui7zvbHbLksK4WKuAtX2vxYWCNJVHFv0rbkbuLPmQ157vfrR3weT2v69Qx1ifl5-hftnDeKdvK4yqhndCAB5KQL9k6921HA2QTMt77xiuzI+BllBV3KAgMkddLDchonHQOD8cqjQkNIWQb80FIOoXgGcRJiAkPHtMWBDCZ6p0ypAxe6EABKngegEQTL2WAhNODcE3jTIGFAobIFeqgmmrUnaHWwRkEhfBAzCOwAmQhyhvTcDvgNch0tAy0JkbxYGYMFFbVYDOQxrB-7TEEd0dwuiyBiIkT6W8s8FhJGdEWGUPQAC0oVsAIDIB9QYkBBHIE2HUVxOiEzIF6PJJYiDZHyNeqIBs3ARjvyAqIaBiDcIMDyawDw0w6DqJKe-JJ7jRHiMkawNJSxpgAH5YnOASWQBpIjnBtM0Lk-J0wtC1MQVPdeu5ehdB6JAKmhSbZFy6VM8mcTen9I8cgH8vYrBKjIEMsgIzKmiAppMMZEz36AKqMA6eOc5mQ0WXQoCKzvwyXPvcjeGzWCJKEY05wuz9nKiOScjwZyLmQHGWxBYdS6GJ2thvR5CzSk206ZABF6yem-L6f8gZqT0nDIwvkiFlyYUqDhTTDh8CHnYDcRhdF1LZDk1uOcAA0qtbYZBqA6gbD8HQDZ2V-CWZOSAXSfl-Lcfi0FxLTkYSFWM1Qep5nhDJd2aB3DlABPcHgYJ8AwkUAie4aJ2RunxJxVslJQKDm9HUYS9I78sk2JyeoipBSXk20BECSlytXk-BpNUq5dDfhiu+H8aF6rEHnGRXeD1k5wKdDpfMuJBARW7l+H8LplxyVAglbiqV2yZVutEMbBNebLWAuwHsm19qwWiBTZMdR4b1HYGVfATVwd26jUcQreSmSrFyOdXYktqLJwznUT2hSXrK4nV7DSVyE9y14sLfJEtfieHZU+mU9RYNhIMSAju9yTlVI20PSVcKp6c1Qxmg1aZkpD0Nlegeq9YM9akwNhvAujaX2T03F8mZBdpyHpuZ8u9e5APnkPZinOX7D1MrKjB-u37uxgwrV4lpvQv0dq3TbatyoxFTN8fQdRsaaYhq6a2-UUKSNRuRQAMX0FUVNcbdwJuRSmzNYhHm9AY0x6Yu7cPAs8bAQjDbqNXsoz0bDMSgJ4eEyBu52AiM1KvaR5W5HO7YEQBE6ASMI3ljU4RaNSbIa8e4mmyUbGTNmrwJxlg3GzOMMgAAMgwoJg5YiFMARwL0MT+mgTo20wKbAHaVqfW44Zx8aLiOqdKQO3CXSsn+euJFvIilj0JfBse8T3ZUuQHPWqTLcM1Q5YM4gm9c0iuI1K0CPLj67GZdsRdLYyWFh5bfQacmyLzmhqpq1lQeW1m0vpXbLpsmhMEb-Upu2-XlB5a82vLr1nRtucnHJzzHzFNEdm3HRB0GkXWYbKO3c6L7Mmd6A2SYIadt5fg0t+ll3Q3H3LOyzliBuW8v5fKtlwqWOSlDWdtxF2FVQqVfqdR4QavXEk+20ArEgA)

In my proposal, the following code:
```typescript
const Category = z.label('Cat', z.object({
  name: z.string(),
  subcategories: z.ref('Cat').array(),
}));
```
would be the [dry](https://en.wikipedia.org/wiki/Don't_repeat_yourself) alternative to this code:
```typescript
interface Category {
  name: string;
  subcategories: Category[];
}

// cast to z.ZodType<Category>
const Category: z.ZodType<Category> = z.lazy(() =>
  z.object({
    name: z.string(),
    subcategories: z.array(Category),
  })
);
```

> I'm pretty confident this is not possible atm.

I hope to have proven you wrong and look forward to further collaboration :^)

**Comment by m93a** [comment]
I think I found something even better. It's possible to have static inference even with the `z.lazy` pattern, if we modify it slightly. The following code would do the same as the two snippets in my previous comment:
```typescript
const Category = z.lazy(self => z.object({
  name: z.string(),
  subcategories: self.array(),
}));
```
[**See it on the TS playground!**](https://www.typescriptlang.org/play?ts=4.9.4#code/LAKA9GAEDuD2BOBrAzgQlAY1gO2QF0gGEBDPAUwHMEBPSAXkgC8A6AG2MeoApkzWAzegD4mzWACMAVmQx4uAb1CRlkbMQC2ZAFyj88AJbYKXAJQAaJSuQBXcRlKUE+ssh28BzYvHjFu50AC+JiYA3KDhIKAQkHgAFmQx1AAOCfrIkPBkxKystHEJmRjW8Mj6AG4JOGToIFi4BBg6LIb8ZPAAPHjJZLCCJORU8NQiDIogKqoa2pAARIiwsKyQiA6D1DMW41a29gNOLjoA2pYTyvKTmjozNnakG5A3u44GB5CHALqQAZunKudql1m-AW4i890eq32rkgABYAExfMyQaJteAIVCQLjEdL6AjIWKwaysAAmkHEZBMJ0g702ATCkQZESZjJZUSg0FipEgwJysGgOPSxEg6kM+nU2TSeH0GAyZDFSVYZE02DwpH0OEgvQeZAqPiWAC1YKTNHEjcg2ZBsTA+EsrflIEk0UlYLxSSx2OTbdg3cxMoITQTieaGcSZOxMhcXEliBgEoxIGMJoZyPB+DGEoRYOp1DgALJkU1BhNUiZeHx+HQAQW8vgAyhh4uLOrE0kJ6b9lAh2gAVSBkAAe5G96RY9cbxA+Qi4zBnsHy8B03ZMOgAqth1dgx0riO1DnE0kiZ8xu+82yWVMRvT2+4OyMPIFvxZPp7P54vl5AAJIqtq8WQbx8dz3FtkEPGcTzPLZlACCIJgHZ14AIZM2jTWMHzwAwjEAm8hyLTNszzAtA3SRMOy6FIdAAcj0QwKEo9sVBgqCbwQhocHwB4MNonRTGEdDMIoQD6SpeCECQn9U3TSAADlrHUcl4GwgdcPSfCc2wfNCxI89lHI6ZKOwOSFPoqkmLg-tWMgOoOMM+S2h4kw+NkuzFIbbdhOY0TEMgZDJLQgAhfQKGTJTb3vNTCK04tmImPSqPEILkxM5izJULy2PqMlEpVBy+MC4KVSE2C0ossSfIk1CEn8hZFUvUKVKILN1M04joo7RIKMgSjxBqrJsGSiZUuUdKrPYggesWPrcroERqsmuq3PFDzzMs3zKofah5MWeq7zwprIta0jfjirrkE2nrWAGxiRNK7zrLxc7FmmkRa0e1gioZWLukgAAFAwRSlCp6E4gTIAAH1UIy2nBh43phhKCoICGJtq7AYcMnIYesb0yH4QwyGJBjhtu8SU3WgAZXE2myQD2gANRw3b0j+sVcXKMgRGUpnGoIjSiLNNqyO6KjWCpvUrt+MpsmsaY6aJr4bss+7IFFlNsh0a8ufvFmAfZqcpdYGX3z4ym1fexad27SDFbKtapOrctaatxnwv2vmoqO04TsostfAl04+CVO88EXeWhpYsrld96gNatrhA+VEPICXPiHbrC2e2tzySfKsmpLXACM97LWi0vahJxdvbeZagXPa+zrKOxjd-YmWAkildjQ9Mm27rGyAm5wWPK8FbBy9PF8xHbjdoRTmbIALnAnazlbbYqqTvxTP8O83Ivh8tUeK5L1S3Zros65Ub3fK35v5YmUNxWHLuUp7jKOKvmRt6Ho-97Hqcj3vy8QZjZzw3r+D+hdxyZ2WiVVaa80IAHkpAfydnvCQ0hZCczClXZq-Mz46Q6vpQoCBiQtysJyTq3Yw4v1GplNBH8v5YPSHQjBPByHTFniIRB6C8BL2gcTWBec0IACVcZtDvLGbssBybEE9JrRhINaIwzOltVgmCGoRXdodfB3s-RiOwLGUhygPR8CfoNahys-QMIajRIwSi3pTmMawYBIgRGtEyPosgkjpGyKtnw5EUBHQFilG0AAtEFbACAyBUj0pAERyBFgVFcXo2MyB2jSSmHvGxFA7EqKRLWPgghv6ASRPAve2EGD5IECMfBpTv5JPcRIqRMi+BpKmCIAA-LElwCSyD1PES4Vpmg8kFJEFoGpe8F470gS0aGDMikWwrp0yZtM4k9L6R45A35uzWAVGQQZZBhkCCRHTIQozxnf1ASUcBi8M4zPgJAOZ8jAKLK-BJa+NzIGrNYIk0RDSXBbJ2YqfZhz+DHNOZAMZMUVC1PkWnagtM7kPLKRbDpkA4UrO6d83pvz+mpPSUMh8BSwVnKhcoGFDUuHINudgNxD5UWUtkLTc4hwADS5VlhkGoFqWs7wdC1lZZ8eZ45ICdK+T8txuLgWEqOQ+AVozVA6mhgEEl7V4F+OiIEvAwT4BhIoBEzI0TvpiuxRKjZALdntHwfihI38sk5Iuj8DslTClPIto6345LubPPeFSappLICevvB8EVbxPiQtVXvQ4iLi6uvHGBZgiK4kECFduD4nxOnHH9RMY16yUlSudUiK28ac04rNdgbZFrrUgqREmoQ+Cw34OwIq+AfiRrK0cRraSmSuK2Ihsoh1ydkXjinPgjtMl3Ve2FsnKkjk54ltNXm6Shbl4wMjn3dgnArFerdTJbtoM+1wwYP2xYU5+DYB4u4fgOhc2eKabI6SQhZ0iA4V0+JWKb14uXa2nOMTyn4IhrWHtgkLb-pklDVy45QP5RCiB-1EM5qo0AqB16KikNwcgKbam5tIFlzreh5ZGcy6TlA5c95UztztCI6eUD6LCOjzw+1CGDKeF0eGKBm9XjmmsEo-Rvxv6LYVsVJIyZmd6D4JjQ1YNnSm26gheJyNiKABi+gSjJtjdueNiaXB4HTaIO57RlOqZEABgTgLb0idrXJ-1Mm2h8e+oBQTt7SPXOwKJug8nv5Sf7tgRAEToBo3DR2CT3Mo00uhoZjiKbxSabC-cpNunmixYMyp-AIgABkD5TO7Mkc5-8OB2iWcC78bGvm+TYG-ZZGJ+ngv3kAn69qNWiyAYEthTpdqiunEa+kZyClWuQxclZhre9oOFQtiGhGyZBtBeG71BawrOkoz6lN34XWNqobG21uGHWJircw3qBFsWHmooZttlQq2CPTMO87TpDmzPCfXPl52p3lCrdy9vA7tLruZfHI5nLbyXOiee4O7+tHLu0trEO7cqLEtuPaLWIQwagereYx96G8OQ3n1+Ky9liBOXct5bKllgr1PihDTDtocO5UQoVbqfBARlunBsy2wIoAgA)

The key here is the `self` argument, which does all the heavy lifting. While at runtime it would hold the exact same value as the `Category` variable, in the TypeScript's type system it has a special non-transparent type, so that it's still possible to statically infer the return type of the lambda.

**Comment by JacobWeisenburger** [comment]
I think this is a really cool idea. Do you think you would be able to make a PR?

**Comment by m93a** [comment]
I started working on it, but then got stuck on the fact that Zod types have both a `_def` property and `In` & `Out` type parameters, ie. the actual resulting type is described in multiple places. I finished the transformer for `_def`, but I haven't implemented the transformers for `In` & `Out` yet.

However, I realized that this makes Zod quite cumbersome for my original use case¹ – therefore I won't be using Zod in my codebase and **cannot guarantee I'll find time to finish the PR**. If anybody wants to tackle this in the meantime, [here's my work in progress](https://github.com/m93a/zod/commit/4cc8076f1ec6503d12e556e7c63b3e3078916d7c).

¹) I need to do a lot of deep mapping on the runtime types, and this doesn't go well with the fact that Zod schemas have three different places where they store type information (`_def`, `In`, `Out`).
